In [113]:
import csv

data_collection = []
x_label = ''
y_label = ''
central_stmt = "{} mean is {} and median is {}."
deviation_stmt = "{} variance is {} and standard deviation is {}."
co_measures_stmt = "{} and {} covariance is {} and correlation is {}."

with open('zadanie_korelacja_source.csv') as source_file:
    source_data = csv.reader(source_file, delimiter=',')
        
    for row in source_data:
        try:
            x = float(row[1])
            y = float(row[2])
        except:            
            x_label = row[1]
            y_label = row[2]
        else:
            r = row[0]
            x = float(row[1])
            y = float(row[2])
            z = [r, x, y]
            data_collection.append(z)
            

def mean(input_list, column):
    result = 0
    collection = []
    
    for row in input_list:
        collection.append(row[column])
            
    result = sum(collection) / len(collection)
    
    return result


def median(input_list, column):
    result = 0
    collection = []
    input_parity = 0
    sample_size = 0
    
    for row in input_list:
        collection.append(row[column])
    
    collection.sort()    
    input_parity = len(collection) % 2
    sample_size = len(collection)
    
    if input_parity == 0:
        a = int(sample_size / 2 - 1)
        b = int(sample_size / 2)
        result = (collection[a] + collection[b]) / 2
    else:
        c = int(sample_size / 2 + 0.5)
        result = collection[c]
        
    return result


def mode(input_list, column):
    print('tu będxzie funkcjsa mody')


def variance(input_list, column):
    collection = []
    deviances = []
    result = 0
    sample_mean = 0
    
    for row in input_list:
        collection.append(row[column])
        
    sample_mean = sum(collection) / len(collection)
    
    for i in collection:
        a = (i - sample_mean) ** 2
        deviances.append(a)
            
    result = sum(deviances) / (len(collection) - 1)
    
    return result


def st_dev(input_list, column):
    result = 0
    
    result = variance(input_list, column) ** (1/2)
    
    return result


def covariance(input_list):
    x_mean = mean(input_list, 1)
    y_mean = mean(input_list, 2)
    deviances = []
    result = 0
    
    for i in input_list:
        a = (i[1] - x_mean) * (i[2] - y_mean)
        deviances.append(a)
        
    result = sum(deviances) / (len(deviances) - 1)
       
    return result


def correlation(input_list):
    result = 0
    
    result = covariance(input_list) / (st_dev(input_list, 1) * st_dev(input_list, 2))
    
    return result


def conf_interval(input_list, column, pop_st_dev, level):
    collection = []
    z = 0
    factor = 0
    result = ()
    sample_mean = 0
    
    for row in input_list:
        collection.append(row[column])
        
    sample_mean = sum(collection) / len(collection)
    
    if level == 99:
        z = 2.576
    elif level == 98:
        z = 2.326
    elif level == 95:
        z = 1.96
    elif level == 90:
        z = 1.645
    else:
        print("Wrong confidence level provided")
        
    factor = z * (pop_st_dev / (len(data_collection) ** (1/2)))
        
    result = ((sample_mean - factor), (sample_mean + factor))
    
    return result    


print(central_stmt.format(y_label, "{:.2f}".format(mean(data_collection, 2)),
                          "{:.2f}".format(median(data_collection, 2))))                          
print(deviation_stmt.format(y_label, "{:.2f}".format(variance(data_collection, 2)),
                            "{:.2f}".format(st_dev(data_collection, 2))))
print(co_measures_stmt.format(x_label, y_label, "{:.2f}".format(covariance(data_collection)),
                              "{:.2f}".format(correlation(data_collection))))
given_conf_int = conf_interval(data_collection, 2, 9.09, 95)
print("{} 95% confidence interval for its mean is from {} to {}.".format(y_label,
                                                                     "{:.2f}".format(given_conf_int[0]),
                                                                     "{:.2f}".format(given_conf_int[1])))


Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.


# Statistics for data provided

Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.

Considering above figures, very strong negative correlation between Fertility rate and Life expectancy at birth has been observed. Such a correlation could be explained by a level of well-being in the countries observed, but further investigation to confirm this hypothesis is needed.
